src/pudl/dfc.py

```diff
@@ -0,0 +1,178 @@
+        """Returns filename where the given dataframe shoudl be stored."""
+        return f"{self._storage_path}/{table_id.hex}/{name}"
+
+    def get(self, name: str) -> pd.DataFrame:
+        """Returns the content of the named dataframe."""
+        try:
+            return pd.read_pickle(self._get_filename(name, self._table_ids[name]))
+        except KeyError:
+            raise KeyError(f"Table {name} not found in the collection.")
+
+    def store(self, name: str, data: pd.DataFrame):
+        """Adds named dataframe to collection and stores its contents on disk."""
+        filename = self._get_filename(name, self._instance_id)
+        if not filename.startswith("gs://"):
+            # Do not make directories when dealing with remote storage.
+            # TODO(rousik): this is fairly crude solution and won't work
+            # for non gcs remote storage.
+            Path(filename).parent.mkdir(exist_ok=True, parents=True)
+        data.to_pickle(filename)
+        self._table_ids[name] = self._instance_id
+
+    def add_reference(self, name: str, table_id: uuid.UUID):
+        """Adds reference to a named dataframe to this collection.
+
+        This assumes that the data is already present on disk.
+        """
+        self._table_ids[name] = table_id
+
+    def __getitem__(self, name: str) -> pd.DataFrame:
+        """Allows accessing dataframes via self[name]."""
+        return self.get(name)
+
+    def __setitem__(self, name: str, data: pd.DataFrame):
+        """Allows adding dataframes via self[name] = value."""
+        return self.store(name, data)
+
+    def __len__(self):
+        """Returns number of tables that are stored in this DataFrameCollection."""
+        return len(self._table_ids)
+
+    def items(self) -> Iterator[Tuple[str, pd.DataFrame]]:
+        """Iterates over table names and the corresponding pd.DataFrame objects."""
+        for name in self.get_table_names():
+            yield (name, self.get(name))
+
+    def get_table_names(self) -> List[str]:
+        """Returns sorted list of dataframes that are contained in this collection."""
+        return sorted(set(self._table_ids))
+
+    def get_table_ids(self) -> Dict[str, uuid.UUID]:
+        """Returns dict mapping dataframe names to their uuid identifiers."""
+        return dict(self._table_ids)
+
+    @staticmethod
+    def from_dict(d: Dict[str, pd.DataFrame]):
+        """Constructs new DataFrameCollection from dataframe dictionary."""
+        return DataFrameCollection(**d)
+
+    def to_dict(self) -> Dict[str, pd.DataFrame]:
+        """Loads the entire collection to memory as a dictionary."""
+        return dict(self.items())
+
+    def update(self, other):
+        """Adds references to tables from the other DataFrameCollection."""
+        # TODO(rousik): typecheck other?
+        for name, table_id in other.get_table_ids().items():
+            self.add_reference(name, table_id)
```

Comment:
This will clobber names which are already present in self, if they're also present in other, in addition to adding names that were not present, right?